### ISSUE TITLE
Unable to deserialize nested collections

### ISSUE THREAD

**Issue by adrien-ben** [cbeust/klaxon#105]
Hi, 

Trying to deserialize a json containing nested collections raises a java.lang.ClassCastException with the following error message : 

`java.base/sun.reflect.generics.reflectiveObjects.ParameterizedTypeImpl cannot be cast to java.base/java.lang.Class`

Here is a code sample to reproduce the issue

```kotlin
class RootListOfMaps(
        val listOfMaps: List<Map<String, Int>>
)

Klaxon().parse<RootListOfMaps>("""
        {
            "listOfMaps": [
                {
                    "Key0": 0,
                    "Key1": 1
                }
            ]
        }
    """)
```

**Comment by cbeust** [comment]
Not sure if you're the same person I was talking to on Slack in #klaxon but yes, I'm working on this issue. Thanks for filing this.

**Comment by adrien-ben** [comment]
It wasn't me. Thanks for looking into it !

**Comment by adrien-ben** [comment]
Hi, it seem's to work just fine now. Thank you.

I spotted another problem though. I don't know if it was introduced with 2.1.6 since i skipped 2.1.5 (or if it relly is an issue).

As workaround for the nested collection problem I created a map delegate :

```kotlin
class MapDelegate(map: Map<String, Int>) : Map<String, Int> by map
```

and replaced 

```kotlin
class Thing(val prop: List<Map<String, Int>>)
```

by

```kotlin
class Thing(val prop: List<MapDelegate>)
```

also I added a converter for the `MapDelegate` class :

```kotlin
val morphTargetConverter = object : Converter<MorphTargetRaw> { ... }
```

And all this worked great. Now, since I upgraded to 2.1.6, the converter was not only applied on `MapDelegate` but also on properties whose type is `Any`.

Sorry, that might not be super clear. I will take some time this evening or tomorrow morning to write some test case reproducing the problem. This is not really an issue for me anymore since deserializing `List<Map<String, Int>>`  works for me now, but wanted to notify you on the potential issue.

**Comment by adrien-ben** [comment]
Hi, here is the test case :

```kotlin
class MapDelegate(map: Map<String, Int>) : Map<String, Int> by map

class Root(val extras: Any?)

val mapDelegateConverter = object : Converter<MapDelegate> {
    override fun fromJson(jv: JsonValue) = MapDelegate(jv.obj?.mapValues { (_, value) -> value as Int } ?: emptyMap())
    override fun toJson(value: MapDelegate) = throw NotImplementedError()
}

class Test {

    @Test
    fun testConverter() {
        val root = Klaxon()
                .converter(mapDelegateConverter)
                .parse<Root>(
                        """{
                            "extras":{
                                "exporterVersion":"2.2.0"
                            }
                        }""")
    }
}
```
It throws a `java.lang.ClassCastException` because it tries to cast `"exporterVersion":"2.2.0"` as an Int.

Do you think I should open another issue ?

**Comment by cbeust** [comment]
I don't think so, that should be enough for me to reproduce the issue. I'll get back to you.

**Comment by cbeust** [comment]
Well, you are doing the cast to an `Int` while it's a string in the JSON document, or am I missing something?

**Comment by adrien-ben** [comment]
Yes, it make sense that it crashes when trying to deserialize `extras` using `mapDelegateConverter`. But I don't understand why this field is being deserialized using `mapDelegateConverter` when its type is `Any` and not `MapDelegate`.

**Comment by cbeust** [comment]
To locate the right converter, Klaxon introspects the type at runtime, its declared type doesn't matter... (otherwise polymorphism would never work).

**Comment by cbeust** [comment]
Never mind, my explanation above is incorrect. Leaving it to avoid confusion, I'm looking into it.

**Comment by cbeust** [comment]
Yes so what's going on is simple: when Klaxon comes across your `Any` (which is `Object`), it goes through all the converters that would produce an object which can be assigned to that `Any`, and obviously, all the converters would work, so it picks the first one it finds (which is your `MapDelegate`).

Not sure the best way to handle this case. Basically, if you declare an `Any` in your object graph, Klaxon can't really do much. Do you know what other JSON libraries do?

**Comment by adrien-ben** [comment]
Thanks for the explanation.

Well, I've tried with [GSON](https://github.com/google/gson) and [Jackson](https://github.com/FasterXML/jackson-databind). And both have similar default behavior. 

GSON maps numbers to `Double`s, booleans to `Boolean`s and strings to `String`s. For complex objects it creates a `LinkedTreeMap<String, Object>` and applies the same rule to each value.

Jaskson does almost the same thing except it also maps integers to `Integer` and the map implementation is `LinkedHashMap`.

I did the tests with the last versions of each libraries.

**Comment by adrien-ben** [comment]
Except for the complex objects part, Klaxon seems to behave is a similar way as Jackson when deserializing fields into Any/Object. Am I right ?

**Comment by cbeust** [comment]
Klaxon will now display a warning if it finds more than one type converter for a value.

**Comment by dvdsgl** [comment]
I'm still unable to parse nested collections. I've [asked how to do this on StackOverflow](https://stackoverflow.com/questions/49163896/parse-json-containing-nested-lists-in-kotlin-with-klaxon). Is there a plan to support this without custom converters?

**Comment by cbeust** [comment]
@dvdsgl Fixed in 2.1.11.

**Comment by dvdsgl** [comment]
Thanks!
